contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+            emit RewardDistributed(operatorAddress, msg.value);
+        } catch Error(string memory reason) {
+            emit RewardDistributeFailed(operatorAddress, bytes(reason));
+        } catch (bytes memory lowLevelData) {
+            emit RewardDistributeFailed(operatorAddress, lowLevelData);
+        }
+    }
+
+    /**
+     * @dev Get new eligible validators from consensus engine
+     */
+    function updateEligibleValidatorSet(
+        address[] calldata validators,
+        uint64[] calldata votingPowers,
+        bytes[] calldata voteAddrs
+    ) external onlyCoinbase onlyZeroGasPrice {
+        uint256 _length = validators.length;
+        if (_length == 0) {
+            return;
+        }
+
+        IBSCValidatorSet.Validator[] memory bscValidatorSet = new IBSCValidatorSet.Validator[](_length);
+        for (uint256 i; i < _length; ++i) {
+            bscValidatorSet[i] = IBSCValidatorSet.Validator({
+                consensusAddress: validators[i],
+                feeAddress: payable(address(0)),
+                BBCFeeAddress: address(0),
+                votingPower: votingPowers[i],
+                jailed: false,
+                incoming: 0
+            });
+        }
+
+        // update validator set in BSCValidatorSet contract
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).updateValidatorSetV2(bscValidatorSet, voteAddrs);
+    }
+
+    function downtimeSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(downtimeSlashAmount);
+        _jailValidator(valInfo);
+
+        uint256 jailUntil = block.timestamp + downtimeJailTime;
+        if (valInfo.jailUntil < jailUntil) {
+            valInfo.jailUntil = jailUntil;
+        }
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, uint248(block.number), SlashType.DownTime);
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    function maliciousVoteSlash(bytes calldata _voteAddr, uint256 height) external onlySlash {
+        address operatorAddress = _voteToOperator[_voteAddr];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(doubleSignSlashAmount);
+        _jailValidator(valInfo);
+
+        SlashRecord storage record = _updateSlashRecord(operatorAddress, slashAmount, height, SlashType.MaliciousVote);
```

Comment:
It is quite strange to return a `storage` type, according to me, the variable `record` is useless here. Eventhough, we use it in the following code, but all of them can be replaced.

Reply:
Done